### PR TITLE
Add edge padding mode

### PR DESCRIPTION
**Pull request by eddieliao**
Adds support for `edge` mode in the ONNX Pad operator.

Closes https://github.com/ROCm/AMDMIGraphX/issues/4069.

### CODE REVIEW

File: test/onnx/verify/pad_edge_1d_test.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+#include <migraphx/register_target.hpp>
+#include <migraphx/verify.hpp>
+#include <onnx_test.hpp>
+
+TEST_CASE(pad_edge_1d_test)
+{
+    migraphx::program p = read_onnx("pad_edge_1d_test.onnx");
+    migraphx::compile_options options;
```
[the remark this answers is not in the document]
Comment (eddieliao) [reply test/onnx/verify/pad_edge_1d_test.cpp]: Removed

File: test/onnx/verify/pad_edge_2d_test.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+#include <migraphx/register_target.hpp>
+#include <migraphx/verify.hpp>
+#include <onnx_test.hpp>
+
+TEST_CASE(pad_edge_2d_test)
+{
+    migraphx::program p = read_onnx("pad_edge_2d_test.onnx");
+    migraphx::compile_options options;
```
[the remark this answers is not in the document]
Comment (eddieliao) [reply test/onnx/verify/pad_edge_2d_test.cpp]: Removed

File: src/onnx/parse_pad.cpp [hunk 3]
```diff
@@ -263,6 +315,11 @@ struct parse_pad : op_parser<parse_pad>
             return reflect_pad(info, pads, args.front());
         }
 
+        if(mode == "edge")
+        {
```
[the remark this answers is not in the document]
Comment (eddieliao) [reply src/onnx/parse_pad.cpp:325]: Added

**Comment by eddieliao** [comment]
Performance tests seem to be normal now after re-running; CI should be passing although the MLIR Debug check on GitHub doesn't seem to reflect what's seen in Jenkins.

File: src/onnx/parse_pad.cpp [hunk 4]
```diff
@@ -111,6 +111,58 @@ static instruction_ref reflect_pad(const onnx_parser::node_info& info,
...
+        std::vector<int64_t> ends(dims.begin(), dims.end());
+        std::vector<instruction_ref> slices;
+
+        auto starts_it = starts.begin() + i;
+        auto ends_it   = ends.begin() + i;
+        auto dims_it   = dims.begin() + i;
+
+        for(int64_t j = 0; j < lcount; j++)
+        {
+            *starts_it = 0;
```
Comment (kahmed10) [inline src/onnx/parse_pad.cpp]: `starts_it` and `ends_it` here don't seem to be impacted by the inner for loop. Or maybe I am missing something?

Comment (eddieliao) [reply src/onnx/parse_pad.cpp]: You're right, this was left over from when I based the implementation off of `reflect` padding. Moved these outside the loop along with the slice instructions to reduce the amount of slices needed overall.

**Comment by eddieliao** [comment]
> should we have more unit tests for 3d, since this is the general method for edge padding.

Added another test with `axes` set, let me know if there are any other tests I should add.

File: src/onnx/parse_pad.cpp [hunk 5]
```diff
@@ -111,6 +111,63 @@ static instruction_ref reflect_pad(const onnx_parser::node_info& info,
     return input;
 }
 
+static instruction_ref edge_pad(const onnx_parser::node_info& info,
+                                const std::vector<int64_t>& pads,
+                                instruction_ref input)
+{
+    size_t num_dims = pads.size() / 2;
```
Comment (kahmed10) [inline src/onnx/parse_pad.cpp]: since we have the input, we should probably assert `num_dims * 2` == `pads.size()` instead

Comment (eddieliao) [reply src/onnx/parse_pad.cpp]: Changed to get `num_dims` from input and added corresponding assert, thanks.

**Review by kahmed10 — APPROVED** [review]

**Review by kentqian — APPROVED** [review]
